### PR TITLE
Porting System.Private.Uri tests

### LINKED ISSUE #6580: [ToFDisable] Port URI tests
As part of moving out of ToF and into the open, I've been going through the "Tree of the future" tests and calculating which of those tests provide coverage that we're missing in the open tests. The goal of this is to get the open tests universally better than the ToF tests so that we can turn the ToF tests off for ProjectK runs and save us all a lot of headache diagnosing test failures from potentially outdated tests.

For System.Runtime, we're missing 2020 lines of coverage. Our total line coverage for Open vs. ToF is 69.7%. However, this includes some data points that might not be relevant to the library being tested so the real coverage is likely lower.

To fulfill the uncovered lines we would need to either create new tests for them or port 56 ToF tests to the open. Those tests are: 

- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnDnsSafeHostTest.IdnDnsSafeHost_IdnOffWithBuiltInScheme_Success.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_IncompleteSurrogate.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriFile_ImplicitUncFile_FragmentNotAllowed.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_ReservedCharacters.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.AppxUriValues.SupportStringsOfForm_MsApp.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriMailToTest.UriMailTo_SchemeUserAt_Throws.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndExplicitUncForwardSlashesWithNoUnicode_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareRelativeAndOriginalHttpWithNoUnicodeAndReservedChar_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_EscapedAscii.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriGetComponentsTest.GetComponents_ASCIIHost_LowerCaseResult.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriGetComponentsTest.GetComponents_UnknownScheme_ComponentsUnaffected.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnHostNameValidationTest.IdnHost_FileFqdn_Success.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriGetComponentsTest.GetComponents_UnknownScheme_LocalHostAndPort_Success.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnDnsSafeHostTest.IdnDnsSafeHost_IPv6Host_ScopeIdButNoBrackets.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriAbsoluteEscaping_FullIPv6Uri_NothingEscaped.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareRelativeAndOriginalHttpWithUnicodeAndReservedCharIriOff_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_804110_TryCreateUri_ShouldNotThrowIndexOutOfRange.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests_FullRun.XML
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnDnsSafeHostTest.IdnDnsSafeHost_SingleLabelAllExceptIntranet_Unicode.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndExplicitFileWithNoUnicode_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsAbsolute_ReturnsFullAbsolute.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriAbsoluteEscaping_EscapeBufferRealloc.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_NewRelativeWithKnownSchemeAndQuery_SuccessButNotWellFormed.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_UnicodePlane0.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsFileLikeUri_MissingRootSlash_ThrowsUriFormatException.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_NewAbsoluteUnregisteredAsRelative_Throws.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndExplicitFileWithReservedChar_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndExplicitUncWithUnicodeIriOn_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndExplicitFileWithUnicodeAndReservedCharIriOn_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriMakeRelative_DifferentHost_ReturnsSecondUri.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_AbsoluteWithColonToRelative_AppendsDotSlash.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_NewRelativeWithUnknownSchemeAndQuery_Throws.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsDoubleCharColinChar_ReturnsCharColinChar.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriMailToTest.UriMailTo_SchemeAndBackslash_Throws.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriMakeRelative_ExplicitDosFileVsUncFile_ReturnsSecondUri.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriMailToTest.UriMailTo_EAI_SomeEscaping.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnHostNameValidationTest.IdnHost_GopherFqdn_Success.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseMadeRelativeToLastSlash_ReturnsDotSlash.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnHostNameValidationTest.IdnHost_Internal_Call_ThrowsException.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_RelativeWithPercentAndColon_Failure.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriFile_ImplicitDosFile_QueryNotAllowed.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriFile_ImplicitDosFile_FragmentNotAllowed.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_EscapedReservedCharacters.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriRelativeFileResolutionTest.IriRelativeResolution_CompareImplcitAndOriginalFileWithNoUnicode_AllPropertiesTheSame.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsAuthority_ReturnsBaseSchemePlusAuthority.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriIsWellFormed_AbsoluteWellFormed_Success.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriEscapingTest.UriUnescapeInvalid_AsciiIncompleteUtf8AsciiIriOn_InvalidUtf8LeftAlone.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_ValidateVeryLongInputString_EscapeUriString.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_ValidateVeryLongInputString_EscapeDataString.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsFragment_ReturnsBasePlusFragment.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriRelativeResolutionTest.Uri_Relative_BaseVsQueryAndFragment_ReturnsBaseAuthorityAndPathPlusQueryAndFragment.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriMakeRelative_ImplicitFileDifferentBaseWithColon_ReturnsSecondUri.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.UriIsWellFormedUriStringTest.UriMakeRelative_DifferentPort_ReturnsSecondUri.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_829158_IncorrectNormalization.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IriTest.Iri_IncompleteEscapedCharacter.xml
- [ ] System.Runtime\4.0.0.0\Uri\UriTests\NCLFunctional.UriTest.IdnHostNameValidationTest.IdnHost_UnknownFqdn_Success.xml

Coverage reports, coverage xml files, and detailed json lists of each piece of missing coverage and the corresponding ToF test that fills it are available on request.

### PR DESCRIPTION
**Pull request by CIPop**
Porting existing internal tests to GitHub.
Porting new Unit Tests to GitHub.
Fixing a production-code bug discovered by the UnitTests.
Code format + sanitize

Fixes #6580.

@davidsh @ianhays @stephentoub @weshaggard PTAL

### CODE REVIEW

**Comment by CIPop** [comment]
3 Failures on Ubuntu:
```
Test Result (3 failures / +3)
System.PrivateUri.Tests.UriIpHostTest.UriIPv6Host_EmbeddedIPv4_Success
System.PrivateUri.Tests.UriIpHostTest.UriIPv4Host_BadAddresses_AllFail
System.PrivateUri.Tests.UriIpHostTest.UriIPv6Host_BadAddresses_AllFail
```
All tests failed on Centos. Tracking with #8337

**Comment by ellismg** [comment]
@dotnet-bot test Innerloop CentOS7.1 Debug Build and Test

**Comment by ellismg** [comment]
@dotnet-bot test Innerloop CentOS7.1 Release Build and Test

**Comment by CIPop** [comment]
@ericeil PTAL

File: src/Common/tests/System/ThreadCultureChange.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace System.Common.Tests
+{
+    public class ThreadCultureChange : IDisposable
```
Comment (davidsh) [inline src/Common/tests/System/ThreadCultureChange.cs]: Should add "sealed"

File: src/System.Private.Uri/tests/System.Private.Uri.Tests.builds [hunk 2]
```diff
@@ -2,10 +2,16 @@
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
-    <Project Include="System.Private.Uri.Tests.csproj">
+    <Project Include="FunctionalTests\System.Private.Uri.Tests.csproj">
       <OSGroup>Unix</OSGroup>
     </Project>
-    <Project Include="System.Private.Uri.Tests.csproj">
+    <Project Include="FunctionalTests\System.Private.Uri.Tests.csproj">
```
Comment (davidsh) [inline src/System.Private.Uri/tests/System.Private.Uri.Tests.builds]: Usually we have the CSPROJ file also have the "Functional" or "Unit" name included.  See System.Net.Http test projects for example.


So, "System.Private.Uri.FunctionalTests.csproj" or "System.Private.Uri.UnitTests.csproj".

**Comment by davidsh** [comment]
Add a few comments, otherwise LGTM.

**Comment by CIPop** [comment]
Missed these failures on OSX only:

```
Test Result (3 failures / +3)
System.PrivateUri.Tests.UriIpHostTest.UriIPv4Host_DottedOctalExtraLeadingZeros_Success
System.PrivateUri.Tests.UriIpHostTest.UriIPv4Host_PartialOctal_Success
System.PrivateUri.Tests.UriIpHostTest.UriIPv4Host_DottedOctal_Success
```

**Comment by CIPop** [comment]
System.Net.Http tests failed on Linux: #8366.

**Comment by CIPop** [comment]
Potential infrastructure issue (tests "Aborted"): #8367

**Comment by CIPop** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please

**Comment by CIPop** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by CIPop** [comment]
Test Innerloop Ubuntu14.04 Debug Build and Test please

**Comment by CIPop** [comment]
Test Innerloop Ubuntu14.04 Release Build and Test please

**Comment by CIPop** [comment]
Thanks for the review @davidsh! I've made the changes.

**Comment by ericeil** [comment]
Looks good.  Thanks for adding these (and filing the resulting issues)!

**Comment by CIPop** [comment]
Thanks for the review @ericeil!

**Pull request merged**